### PR TITLE
Add an endpoint to catalog-backend that refreshes a single specified location in the catalog.

### PR DESCRIPTION
**Pull request by ursaregina**
See title.

Signed-off-by: Christina Marie Rahaim <crahaim@wayfair.com>

## Hey, I just made a Pull Request!

Our particular use case: we created a barebones editor in Backstage to allow users to edit entity configuration for certain in-house entity sources in Backstage itself. Hitting this refresh endpoint syncs up changes made to the source of truth for these entities with Backstage, allowing users to see their changes right away.

#### :heavy_check_mark: Checklist

- [Should I?] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [Found none relevant to add to] Added or updated documentation
- [Y] Tests for new functionality and regression tests for bug fixes
- [N/A] Screenshots attached (for UI changes)
- [Y] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/catalog-backend/src/next/DefaultLocationStore.ts [hunk 1]
```diff
@@ -111,6 +111,39 @@ export class DefaultLocationStore implements LocationStore, EntityProvider {
...
+      throw new NotFoundError(`Found no location with type ${type}`);
+    }
+
+    const id = location[0].id;
+
+    const isRefreshRowExists = await this.db<DbRefreshStateRow>('refresh_state')
+      .where('entity_id', id)
+      .select();
+
+    let result = `Location found, refresh row not found for type ${type}, please wait a while and try again`;
```
Comment (ursaregina) [inline plugins/catalog-backend/src/next/DefaultLocationStore.ts:133]: Please advise on messaging here- as I understand it, a row will be inserted eventually/soon if the type is new and doesn't have one yet, but I'm not confident in that assessment.

File: plugins/catalog-backend/src/next/DefaultLocationStore.test.ts [hunk 2]
```diff
@@ -178,4 +178,54 @@ describe('DefaultLocationStore', () => {
...
+        });
+        const result = await store.refreshLocation(type);
+        expect(result).toEqual(
+          `Location found, refresh row not found for type ${type}, please wait a while and try again`,
+        );
+      },
+      60_000,
+    );
+
+    /*    it.each(databases.eachSupportedId())(
```
Comment (ursaregina) [inline plugins/catalog-backend/src/next/DefaultLocationStore.test.ts:213]: I would like to implement this, but I'm a little stumped as to how to mock a refresh row for now and a little time crunched to do it. If required, I'd appreciate some pointers; otherwise happy to remove and/or add a TODO.

File: plugins/catalog-backend/src/next/NextRouter.ts [hunk 3]
```diff
@@ -142,6 +142,11 @@ export async function createNextRouter(
         const output = await locationService.getLocation(id);
         res.status(200).json(output);
       })
+      .get('/locations/:type/refresh', async (req, res) => {
```
Comment (benjdlambert) [inline plugins/catalog-backend/src/next/NextRouter.ts:145]: So i'm not sure about the API and REST endpoint here.

I think this opens up a bigger discussion here to what we want to public API of this to be.

Should be it be a `location` or could it be a single `entityRef` and then it can refresh the parent location?

File: plugins/catalog-backend/src/service/router.ts [hunk 4]
```diff
@@ -173,6 +173,11 @@ export async function createRouter(
         const output = await locationService.getLocation(id);
         res.status(200).json(output);
       })
+      .get('/locations/:type/refresh', async (req, res) => {
```
Comment (benjdlambert) [inline plugins/catalog-backend/src/service/router.ts:176]: This is the old catalog stuff here, so we shouldn't touch this. This will be removed in future versions.

File: plugins/catalog-backend/src/next/DefaultLocationStore.ts [hunk 5]
```diff
@@ -111,6 +111,39 @@ export class DefaultLocationStore implements LocationStore, EntityProvider {
...
+      const refreshResult = await this.db<DbRefreshStateRow>('refresh_state')
+        .update({ next_update_at: Date.now().toString() }, [id])
+        .where('entity_id', id);
+
+      if (refreshResult.length) {
+        result = `Refresh row for type ${type} updated`;
+      }
+    }
+
+    return result;
```
Comment (benjdlambert) [inline plugins/catalog-backend/src/next/DefaultLocationStore.ts:144]: Really not sure about returning a `string` from this really.

Wondering what return type should be here to be honest, think this is going to require some thought and some input from @backstage/maintainers when some of them are back from holiday.

Feels like we could hold on to the connection until the refresh is complete and return the refreshed entity, but that could also open up problems if the refresh takes a while.

**Review by benjdlambert — COMMENTED** [review]
Going to also add the `after-vacation` label here, as I think this requires some more input from other maintainers here as to what we want the API to be.

File: plugins/catalog-backend/src/next/NextRouter.ts (hunk 3 above)
Comment (ursaregina) [reply plugins/catalog-backend/src/next/NextRouter.ts:145]: If we're opening up that question, I would probably say "both", i.e. create both endpoints, unless we can come up with a good reason not to, just for the flexibility.

File: plugins/catalog-backend/src/service/router.ts (hunk 4 above)
Comment (ursaregina) [reply plugins/catalog-backend/src/service/router.ts:176]: Ah, changes I saw must have been from before release, will remove.

File: plugins/catalog-backend/src/next/DefaultLocationStore.ts (hunk 5 above)
Comment (ursaregina) [reply plugins/catalog-backend/src/next/DefaultLocationStore.ts:144]: re: what we're returning, I don't much care about the type but we would like some messaging of some kind.

**Comment by freben** [comment]
Hi! So I've taken a longer look at this now and giving it some thought. It's a pretty important function with broader implications and we want to get it right.

Firstly, it's very unexpected to change from `:id` to `:type` semantics mid-URL and "overlaying" needs. So we should probably move to a different URL for this.

Users will want/need to have things marked dirty in several different ways. The hints about dirty-marking can also come from several different places. For example, it is very reasonable to want to listen for webhook events from the source code management system, and making sure that the catalog immediately refreshes itself after you've pressed merge on a PR that affects a `catalog-info.yaml` file.

## REST

For "manually" triggered refreshes, such as the one you are proposing here, I think we could do something like the following.

Introduce a completely new endpoint at the root, `/refresh`. It is a POST endpoint that accepts a JSON payload. The payload is an object, for example on a form like this:

```json
{
  "action": "refresh",
  "match": [
    {
      "type": "location",
      "location": { "type": "my-custom-type" }
    },
    {
      "type": "entity",
      "entity": { "ref": "component/default:my-component-name" }
    },
    ...
  ]
}
```

this lets you match in any of several ways, and to do it as a batch action. We can introduce the matchers gradually, don't need to invent them all at the same time. But the framework for it needs to be in place.

You asked about the return type, and I am not sure what it should be but maybe it could be also an object that for example lists all entities that it matched? Not sure. I'd prefer to not return anything to begin with, and only introduce a return type when there's a strict business need to do so - if ever. The minute that we introduce a new interface, it amounts to making a long term promise to uphold that interface, and that can be hard to do with very little information at hand.

## Business Logic

Now, regarding the business logic of what to do.

There are at least two things at play here. The `EntityProvider` instances that feed "root" entities into the `refresh_state` table, and processors that then regularly act upon the roots and other things in the state table, eventually turning them into `final_entities`.

What you've seen is the `next_update_at` column, which you can reset. This makes processors re-run over that entity some time soon in the future. That is a good start, and covers a lot of current use cases.

But we'll want to notify the entity providers as well. Users may have for example implemented an LDAP org reader in the form of an entity provider. If they get a notification that a user has updated their profile, they will want to issue a single LDAP read for that user and "seed" that into the database. Which will, in turn, also make processors run over it soon. Interestingly, in that case now that i think about it, we'd actually prefer to not move the `next_update_at` timestamp because then it'll just be unnecessarily re-processed with no changes, then replaced by the provider, and immediately processed again. But ... that's a trickier one, let's not worry about that upfront.

## The Middle Layers

This tells me that there needs to be a "bridge" between the API and the location db + entity providers. A POST to the `/refresh` endpoint should probably be parsed and received by a new `RefreshHandler` class for example, a small proxy which can distribute events to anybody who's subscribed to it. The catalog builder would wire up that handler to the providers so that they can listen for the stream of `match` things from the JSON payload above. And something entirely different will probably have to listen as well, to do the `next_update_at` resets.

## Summary

This introduces a new REST endpoint, a few typescript types, an event bridging class, and a bit of wiring to make existing classes listen to the events from the latter. The listeners will be responsible for acting on the refreshes if they are on a form that they understand, or just ignore them if they don't.

I know this sounds a bit daunting. Let me know how you feel about it. I can maybe help out by getting a framework started for this.

**Comment by ursaregina** [comment]
Hey @freben- thanks for the thoughtful feedback. Definitely more work than expected but I think I'm game ;) I'm replying without quite enough time to do as much research as I'd like since I'm on vacation next week and I want to get back to you before I go. My teammates may continue this discussion or do some work on this in the meantime.

First a set of "yeses":

- I'm good with the new endpoint URL and JSON payload
- Same with the `RefreshHandler`, I think that makes sense
- If you have time to start the matcher framework as (if that's what you meant), that would be excellent, or at least if there's an example of the pattern you're thinking of somewhere that would help tremendously. I have ideas on how I would do it, but not my codebase :)

### Return Type

I get what you're saying about not creating that contract before it's necessary, however returning no information makes the process as we're using it a little opaque from a user perspective. We store most of our entity configuration data in an external database rather than as catalog-info.yaml files in github, and Backstage serves as the UI for editing that data- meaning they have to wait for the processors to run in order for them to see their changes in Backstage. We are using a ten minute refresh interval, which means Backstage can get the changes up to ten minutes later- not a great user experience, and it sometimes obfuscates _actual_ problems with the processors or the entity config database.

To that end, would another endpoint URL or HTTP method on this endpoint URL getting (or GETting) entity data that includes last refresh time make sense? Does a better place for that already exist? (I'll poke around for that a bit, but if you know offhand let me know). That could allow us to do the following:
  - save the entity configuration data
  - send off a refresh request
  - maybe wait a second or two
  - send off a status check request
  - use the results of the status check to inform the user

Still a little clunky, but better than waiting ten minutes before knowing if it worked properly :) Very much open to better ideas as well, I went for simplest solution there.

### Business Logic

Can you try explaining that last paragraph again? I've dug around in the code just a little, but the entity provider code I've seen is pretty abstract and I am running out of time again. Actually, are there design docs for catalog next? I admittedly haven't looked, my apologies.

---

TL;DR: I think with maybe a little guidance we can and want to make this happen. I do hope to address our use case above in a way that is general enough for this codebase, although I'd probably separate that into a separate task given the scope of this one.

**Comment by freben** [comment]
> * If you have time to start the matcher framework as (if that's what you meant), that would be excellent, or at least if there's an example of the pattern you're thinking of somewhere that would help tremendously. I have ideas on how I would do it, but not my codebase :)

Hmmm maybe :) There's a ton going on. I'll update this issue if I do.

> To that end, would another endpoint URL or HTTP method on this endpoint URL getting (or GETting) entity data that includes last refresh time make sense?

There is no last refresh time actually in the table, just the next update time. As a matter of fact, I believe that there's no solid (race-free) way currently to deduce whether certain subgraphs have gone through one more round of complete processing. We currently only "mark" roots for future processing, and that can have implications for arbitrarily deeply nested subtrees, processed by any number of distributed workers. So it's a bit tricky.

Maybe we can do something for a single level. If we do choose to pay the cost of "touching" all entities that are processed, even though they had no changes, then sure we could have some kind of generation marker on each `refresh_state` row.

> Can you try explaining that last paragraph again? I've dug around in the code just a little, but the entity provider code I've seen is pretty abstract and I am running out of time again. Actually, are there design docs for catalog next? I admittedly haven't looked, my apologies.

No worries. Yeah it's not really documented, that's on our todo as well. Apologies that this hasn't landed yet. The providers are actually SUPER simple; they can only tell the rest of the system "here's a new entity that i just discovered", "hey an entity that i used to know of should disappear now", and "here's the entire list of entities that I currently know of". You can send those events at any time. The config entity provider only does a full update at startup. The location entity provider does a full update at startup and little deltas each time you add or remove a location.

What I meant was, one of those providers may be interested in the signal that something that it was responsible for, should be updated. Because if you have a provider that feeds data from ldap, then _that provider_ needs to re-read from ldap into the database. Re-processing what's in the database has no effect whatsoever because no ldap reading would be happening _at the processor_ level in that case.

> TL;DR: I think with maybe a little guidance we can and want to make this happen. I do hope to address our use case above in a way that is general enough for this codebase, although I'd probably separate that into a separate task given the scope of this one.

Yeah I totally understand that. I've turned this into a bit of a thing :)

Just some rough thoughts:

```ts
// in the catalog builder or perhaps packages/backend
const refreshHandler = new RefreshHandler();
const ldapProvider = new LdapProvider();
const databaseRefreshHandler = new DatabaseRefreshHandler(db);
ldapProvider.subscribeRefreshes(refreshHandler);
databaseRefreshHandler.subscribeRefreshes(refreshHandler);

// in the router
.post('/refresh', (req, res) => {
  const r = parseRefreshRequest(req);
  refreshHandler.send(r);
})

// in the refresh handler
subscribe(listener) {
  this.listeners.push(listener)
}
send(r) {
  for (const listener of this.listeners) {
    listener(r);
  }
}
```

eh just brain dumping to get some overall idea about things

**Comment by freben** [comment]
Just wanted to mention that we've started taking a look at this internally and hope to have a design to share some time in the near future.

**Comment by ursaregina** [comment]
That's awesome and would be a tremendous help to get us going!

**Comment by benjdlambert** [comment]
This work has been superceded with a few PR's.

#7186 #7175
